components/authz-service/storage/postgres/migration/sql/61_use_role_db_id_in_iam_statements.up.sql

```diff
@@ -0,0 +1,233 @@
+-- the database for actions (and maybe resources). clean up any such instances.
+UPDATE iam_statements t SET actions = '{}' WHERE actions IS NULL;
+ALTER TABLE iam_statements ALTER COLUMN actions SET NOT NULL;
+UPDATE iam_statements t SET resources = '{}' WHERE resources IS NULL;
+ALTER TABLE iam_statements ALTER COLUMN resources SET NOT NULL;
+
+-- also set policy_id to NOT NULL since a statement can't exist
+-- outside of a policy. there should not be any instances of
+-- orphaned statements, but adding the delete line just in case
+-- so migrations don't explode if there are any somehow (they
+-- are useless if they exist anyway).
+DELETE FROM iam_statements WHERE policy_id IS NULL;
+ALTER TABLE iam_statements ALTER COLUMN policy_id SET NOT NULL;
+
+ALTER TABLE iam_statements ADD COLUMN role_id INTEGER;
+
+-- populate role_id in all statements that have a role
+UPDATE iam_statements t
+  SET role_id = (SELECT db_id FROM iam_roles WHERE id = t.role)
+  WHERE role != '';
+
+ALTER TABLE iam_statements DROP COLUMN role;
+
+ALTER TABLE iam_statements
+    ADD CONSTRAINT iam_statements_role_id_fkey FOREIGN KEY (role_id) REFERENCES iam_roles(db_id);
+
+-- This need to be DEFERRABLE INITIALLY DEFERRED so that we don't get a FKEY error before
+-- the trigger works it's magic below.
+ALTER TABLE iam_statements ALTER CONSTRAINT iam_statements_role_id_fkey DEFERRABLE INITIALLY DEFERRED;
+
+-- i wanted use the above FKEY constraint with ON DELETE to clean up any statement where the
+-- role getting removed would result in a NULL role_id in iam_statements. that is doable, but
+-- what becomes tricky is adding the additional logic to completely remove the statement if the affected
+-- policy would also have no actions. so instead of splitting that logic up, let's handle it all in a trigger.
+
+CREATE OR REPLACE FUNCTION purge_statements_with_no_actions_or_role() RETURNS TRIGGER AS $$
+  BEGIN
+    -- for statements that will still have actions, simply remove the role since those
+    -- statements are still valid
+    UPDATE iam_statements SET role_id=NULL WHERE role_id=OLD.db_id AND actions != '{}';
+
+    -- for statements that become invalid (no role or actions), delete them
+    DELETE FROM iam_statements WHERE role_id=OLD.db_id AND actions = '{}';
+
+    -- if as a result, a policy now has no statements, remove the policy unless it's
+    -- chef-managed (chef-managed case should never happen since chef-managed roles can't be
+    -- deleted, but just to be safe adding it in)
+    DELETE FROM iam_policies USING iam_policies AS p
+      LEFT OUTER JOIN iam_statements AS s ON p.db_id=s.policy_id
+      WHERE s.policy_id IS NULL AND iam_policies.id=p.id AND p.type != 'chef-managed';
+
+    RETURN NULL;
+  END
+$$
+LANGUAGE plpgsql;
+
+CREATE TRIGGER on_role_deletion AFTER DELETE ON iam_roles
+FOR EACH ROW
+EXECUTE PROCEDURE purge_statements_with_no_actions_or_role();
+
+CREATE OR REPLACE FUNCTION role_db_id (_id TEXT)
```

Comment:
❓ Can we use plpgsql and `SELECT db_id INTO STRICT ...`? That way, it'll throw an exception (P0002 IIRC) instead of returning NUL if the `SELECT` returns nothing.

Reply:
Lemme see how much stuff that breaks.